VirtoCommerce.Storefront/Domain/Cart/CartBuilder.cs

```diff
@@ -168,17 +168,26 @@ public virtual Task RemoveItemAsync(string id)
             return Task.FromResult((object)null);
         }
 
-        public virtual Task AddCouponAsync(string couponCode)
+        public virtual Task<bool> AddCouponAsync(string couponCode)
         {
             EnsureCartExists();
-            Cart.Coupon = new Coupon { Code = couponCode };
-            return Task.FromResult((object)null);
+            if (!Cart.Coupons.Any(c => c.Code == couponCode))
+            {
+                Cart.Coupons.Add(new Coupon { Code = couponCode });
+                return Task.FromResult(true);
+            }
+            else
+            {
+                return Task.FromResult(false);
+            }
         }
 
-        public virtual Task RemoveCouponAsync()
+        public virtual Task RemoveCouponAsync(string couponCode)
         {
             EnsureCartExists();
-            Cart.Coupon = null;
+            var coupon = Cart.Coupons.FirstOrDefault(c => c.Code == couponCode);
```

Comment:
You can use the `Cart.Coupons.RemoveAll` instead of these lines

Reply:
Is it ok to convert it using .ToList() first?